src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -422,7 +422,9 @@ internal unsafe void Start()
                         // The base FileSystemWatcher does a match check against the relative path before combining with
                         // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
                         ReadOnlySpan<char> relativePath = ReadOnlySpan<char>.Empty;
-                        if (!path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
+                        if (!path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase)
+                            && path.Length >= _fullDirectory.Length
+                            && _fullDirectory.AsSpan().Equals(path.Slice(0, _fullDirectory.Length), StringComparison.OrdinalIgnoreCase))
```

Comment:
You can use `path.StartsWith(fullDirectory, StringComparison.OrdinalIgnoreCase)` for this.